components/canvas/utils/handlePasteNode.tsx

```diff
@@ -0,0 +1,37 @@
+import { NodeData, NodeDataFull } from "@/types/NodeData";
+import { NodeTypes } from "@/types/NodeTypes";
+import { Position, Node } from "reactflow";
+import { isValidTarget } from "./isValidTarget";
+import { NodeDataApiRequestBody } from "@/types/NodeDataApi";
+import { getNodeTypeName } from "@/utils/getNodeTypeName";
+
+export const handlePasteNode = (
+  source: Node<NodeData>,
+  target: Node<NodeData> | undefined,
+  nodes: Node<NodeDataFull, string | undefined>[],
+  addNode: (
+    parentId: string,
+    data: NodeDataApiRequestBody,
+    position: Position
+  ) => void
+) => {
+  const sourceType = getNodeTypeName(source.data.type);
+  const targetType = getNodeTypeName(target?.data.type);
+
+  if (!target?.id) {
+    throw new Error(
+      "Unable to paste: Hover a card before attempting to paste ⛔"
+    );
+  }
+  if (isValidTarget(source, target, nodes, false)) {
+    addNode(
+      target.id,
+      source.data,
+      source.type === NodeTypes.mainActivity ? Position.Right : Position.Bottom
+    );
+  } else {
+    throw new Error(
+      `Unable to paste: ${sourceType} card is not allowed below ${targetType} card ⛔`
+    );
+  }
+};
```

Comment:
Throwing an error could be an early return to have the actual functionality as the base case. This also removes the need for nesting.

Reply:
Not sure if I completely understand what you mean by having the actual functionality as the base case, but check out the changes i made

Reply:
Your changes looks how I wanted it 👌